### PR TITLE
chore(zipkin): adds span serializer for different formats.

### PR DESCRIPTION
**Pull request by jcchavezs**
Integrating span serializer.

In a next PR I will move all the serialization from the span into the serializators.

Ping @adriancole @dio

### CODE REVIEW

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 1]
```diff
@@ -38,6 +38,22 @@ std::string SpanBuffer::toStringifiedJsonArray() {
...
       stringified_json_array += ",";
       stringified_json_array += span_buffer_[i].toJson(version_);
+          }
+  }
+  stringified_json_array += "]";
+
+  return stringified_json_array;
+}
+
+std::string SpanBuffer::toStringifiedJsonArray(SpanSerializer& span_serialier) {
```
Comment (jcchavezs) [inline source/extensions/tracers/zipkin/span_buffer.cc]: Most likely we delegate the SpanBuffer serialization into the serializer so we avoid spreading this knowledge across Span and SpanBuffer. WDYT @dio?

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 2]
```diff
@@ -18,31 +18,8 @@ bool SpanBuffer::addSpan(const Span& span) {
...
-    const uint64_t size = span_buffer_.size();
-    for (uint64_t i = 1; i < size; i++) {
-      stringified_json_array += ",";
-      stringified_json_array += span_buffer_[i].toJson(version_);
-    }
-  }
-  stringified_json_array += "]";
-
-  return stringified_json_array;
+std::string SpanBuffer::serialize(SpanSerializer& serializer) {
```
Comment (jcchavezs) [inline source/extensions/tracers/zipkin/span_buffer.cc:21]: I was always curious about a good example for [visitor pattern](http://butunclebob.com/ArticleS.UncleBob.IuseVisitor). I think this one is a good one.

File: source/extensions/tracers/zipkin/span_buffer.h [hunk 3]
```diff
@@ -61,19 +62,13 @@ class SpanBuffer {
   uint64_t pendingSpans() { return span_buffer_.size(); }
 
   /**
-   * @return the contents of the buffer as a stringified array of JSONs, where
-   * each JSON in the array corresponds to one Zipkin span.
+   * @return the contents of the buffer as string
    */
-  std::string toStringifiedJsonArray();
+  std::string serialize(SpanSerializer& serializer)
```
Comment (jcchavezs) [inline source/extensions/tracers/zipkin/span_buffer.h:67]: Now SpanBuffer does not have to be aware of proto or json serialization.

File: test/extensions/tracers/zipkin/span_buffer_test.cc [hunk 4]
```diff
@@ -15,31 +15,33 @@ TEST(ZipkinSpanBufferTest, defaultConstructorEndToEnd) {
   SpanBuffer buffer(envoy::config::trace::v2::ZipkinConfig::HTTP_JSON_V1);
 
   EXPECT_EQ(0ULL, buffer.pendingSpans());
-  EXPECT_EQ("[]", buffer.toStringifiedJsonArray());
+  // tests are commented out for now, I will move them to serialization once
+  // have more expertise with tests.
```
Comment (jcchavezs) [inline test/extensions/tracers/zipkin/span_buffer_test.cc:19]: ☝️

File: source/extensions/tracers/zipkin/zipkin_tracer_impl.cc [hunk 5]
```diff
@@ -126,18 +126,75 @@ Tracing::SpanPtr Driver::startSpan(const Tracing::Config& config, Http::HeaderMa
   return active_span;
 }
 
+HTTPSpanSerializerV1::HTTPSpanSerializerV1() {}
```
Comment (jcchavezs) [inline source/extensions/tracers/zipkin/zipkin_tracer_impl.cc:129]: @dio is it OK if I move this into another file? I guess yes (as long as that file is part of the compilation like in C) but I want to double check.

File: source/extensions/tracers/zipkin/span_serializer.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  virtual ~SpanSerializer() {}
+
+  /**
+   * Method that a concrete SpanSerializer class must implement to serialize spans.
+   *
+   * @param span The span that needs action.
+   */
+  virtual std::string serialize(std::vector<Span> spans);
+
+  virtual std::string contentType();
```
Comment (jcchavezs) [inline source/extensions/tracers/zipkin/span_serializer.h:23]: The serializer is the one that knows the content type so it makes sense we get that information from here.

**Comment by dio** [comment]
OK, this is interesting. I have pushed the ["base" PR](https://github.com/envoyproxy/envoy/pull/6985) for us to work on (I'll fix the build problem if that exists).

**Comment by dio** [comment]
This is the corresponding branch in this repo: https://github.com/envoy-zipkin/envoy/tree/zipkin-proto3

**Pull request closed without merging**